lib/natalie/repl.rb

```diff
@@ -14,33 +14,53 @@ def self.readline(prompt, _)
 
 module Natalie
   class Repl
+    def initialize
+      @to_clean_up = []
+      @var_decls = {}
+      # Readline.completion_proc = self.method(:completion)
+    end
+
     def go
-      to_clean_up = []
       env = nil
-      vars = {}
       loop do
-        break unless (line = get_line)
-        ast = Natalie::Parser.new(line, '(repl)').ast
-        last_node = ast.pop
-        ast << last_node.new(:call, nil, 'puts', s(:call, s(:lasgn, :_, last_node), 'inspect'))
-        out = Tempfile.create('natalie.so')
-        to_clean_up << out
-        compiler = Compiler.new(ast, '(repl)')
-        compiler.repl = true
-        compiler.vars = vars
-        compiler.out_path = out.path
-        compiler.compile
-        vars = compiler.context[:vars]
-        lib = Fiddle.dlopen(out.path)
-        env ||= Fiddle::Function.new(lib['build_top_env'], [], Fiddle::TYPE_VOIDP).call
-        eval_func = Fiddle::Function.new(lib['EVAL'], [Fiddle::TYPE_VOIDP], Fiddle::TYPE_VOIDP)
-        eval_func.call(env)
+
+          break unless (line = get_line)
+          out = Tempfile.create('natalie.so')
+          do_compile(line, out) 
+          lib = Fiddle.dlopen(out.path)
+          env ||= Fiddle::Function.new(lib['build_top_env'], [], Fiddle::TYPE_VOIDP).call
+          eval_func = Fiddle::Function.new(lib['EVAL'], [Fiddle::TYPE_VOIDP], Fiddle::TYPE_VOIDP)
+          eval_func.call(env)
       end
-      to_clean_up.each { |f| File.unlink(f.path) }
+    rescue Interrupt => e
+      system('stty', `stty -g`.chomp)
+      exit
+    ensure
+      @to_clean_up.each { |f| File.unlink(f.path) }
     end
 
     private
 
+    def do_compile(line, outfile)
+      ast = Natalie::Parser.new(line, '(repl)').ast
+      last_node = ast.pop
+      ast << last_node.new(:call, nil, 'puts', s(:call, s(:lasgn, :_, last_node), 'inspect'))
+      
+      @to_clean_up << outfile
+      
+      compiler = Compiler.new(ast, '(repl)')
+      compiler.repl = true
+      compiler.vars = @var_decls
+      compiler.out_path = outfile.path
+      compiler.compile
+      
+      @var_decls = compiler.context[:vars]
+    end
+
+    def completion(str)
+      # TODO
+    end
+
```

Comment:
I think I got a sneak peak of something else you're working on. 😉

Reply:
Oops didn't realize the repl stuff got pushed :joy:
Yeah I at least made it not panic on `Ctrl+C` now. I definitely need a little more time with the compiler before I figure out how to make completion work.

Reply:
I pulled out the REPL changes and merged this via the terminal. If you think the REPL stuff is good to go, feel free to make a new PR for that! Can't wait!